### DISCUSSION TITLE
PHP - Creazione Fattura -> Unauthorized

### DISCUSSION THREAD

**Discussion by next731** [fattureincloud/api#72] [Q&A] [answered]
Buongiorno.
Sto implementando la API V2 in PHP (<7) usando cURL come metodo di chiamata in quanto non ho possibilità di installare la SDK (PHP7).
Fino all'autenticazione oAuth, tutto bene, Mi viene restituito tutto ciò che mi serve:

```
Array
(
    [token_type] => bearer
    [access_token] => 'xxxxxx'
    [refresh_token] => 'yyyyyy'
    [expires_in] => 86400
)
```

A questo punto, utilizzando cURL con PHP:

```
$Token = <access_token precedentemente restituito>;
$CompanyID = <Mia Company ID>;


$curlSend = curl_init();
curl_setopt_array($curlSend, array(
  CURLOPT_URL => 'https://api-v2.fattureincloud.it/c/'.$CompanyID.'/issued_documents',
  CURLOPT_RETURNTRANSFER => true,
  CURLOPT_ENCODING => '',
  CURLOPT_MAXREDIRS => 10,
  CURLOPT_TIMEOUT => 0,
  CURLOPT_FOLLOWLOCATION => true,
  CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
  CURLOPT_CUSTOMREQUEST => 'POST',
  CURLOPT_POSTFIELDS => json_encode($aData),
  CURLOPT_HTTPHEADER => array(
    'Accept: application/json',
    'Authorization: Bearer '.$Token
  ),
));

$response = curl_exec($curlSend);

curl_close($curlSend);
echo $response;

```

Il risultato ($response) è:
`{"error":{"message":"Unauthorized."}}`

All'apertura della richiesta oAuth, richiedo uno scope molto ampio - giusto per essere sicuro:

`issued_documents.invoices:a issued_documents.credit_notes:a issued_documents.receipts:a issued_documents.orders:a issued_documents.quotes:a issued_documents.proformas:a issued_documents.delivery_notes:a`

Sono abbastanza certo che basterebbe anche solo **issued_documents.invoices:a**

Dove sto sbagliando?
Grazie mille!

**Comment by valmoz** [comment]
Buongiorno @next731 
prima di tutto, le chiederei di provare di utilizzare il suo access token tramite l'[API Reference](https://developers.fattureincloud.it/api-reference#auth) oppure tramite [Postman](https://developers.fattureincloud.it/docs/quickstarts/postman-collection). Ciò ci permetterà di stabilire se l'access token è stato inizializzato correttamente, oppure se i permessi non sono stati inseriti nel modo giusto all'interno della richiesta OAuth.
Un'altra ipotesi è che si stia facendo riferimento a una company o a una risorsa non appartenenti all'account cui si riferisce l'access token.

**Reply by next731** [reply]
Grazie per la risposta @valmoz. Ho provato con Postman.
Ho copiato l'access-token restituito nel bearer token di Postman (1). Ho eseguito il codice (ovviamente con la company ID corretta) e mi ha restituito un errore (2) che è assolutamente corretto e gestibile.

A questo punto, per non saper ne leggere, ne scrivere, ho copiato il codice autogenerato di Postman (3) nello script PHP, ed utilizzando l'access-token corretto, continuo a ricevere: 
`{"error":{"message":"Unauthorized."}}`

**Reply by next731** [reply]
Una curiosità: nel log dell'applicazione c'è solo la richiesta effettuata da Postman
![image](https://user-images.githubusercontent.com/10614002/190162187-1251f8bc-8b64-4b6e-9a27-d7225caa642c.png)

**Reply by valmoz** [reply]
Le chiederei di controllare quali sono le differenze tra la chiamata che sta effettuando e quella utilizzata tramite Postman. Sta usando parametri diversi in Postman? (body, companyID o access token)?

A noi risulta che il codice da lei inviato funzioni correttamente. Le mie ipotesi sono le seguenti:

- il companyID e l'access token non combaciano o non sono passati correttamente (se un access token ha accesso alla company 11111 e io provo ad utilizzare /c/22222/issued_documents otterrò un errore)
- Il body della richiesta fa riferimento ad altre risorse (ad esempio payment account e payment method) che non appartengono alla sua company?

Le chiederei di provare lato codice esattamente gli stessi parametri e body JSON utilizzati in postman, dovrebbe ricevere un 422 anche in questo caso.

**Reply by next731** [reply]
@valmoz ,Come già scritto, purtroppo, ho fatto COPIA/INCOLLA ASSOLUTAMENTE STATICO della parte (3) di Postman in un file PHP e l'ho lanciato.
Il risultato è sempre lo stesso.

Se mi date una mail (anche privatamente), posso condividerVi lo script con tutti i valori (Company e Token) usati in Postman (rigenerati e funzionanti entro il limite predefinito).


Grazie.

**Reply by valmoz** [reply]
Buongiorno,
può contattare il supporto clienti sulla pagina FIC allegando la discussione corrente, in modo tale da poter inviare tutti i dati sensibili e il codice.

P.S.: cancello l'immagine relativa a postman perché contiene una parte dell'access token

**Comment by next731** [comment, marked as answer]
@valmoz , dopo svariati tentativi ho trovato l'inghippo.
Nel setup default di Postman, cURL viene configurato con questo parametro:

`CURLOPT_FOLLOWLOCATION => true`

Deve essere omesso, ovvero, lasciato come da default FALSE.

Grazie per l'attenzione.

**Reply by valmoz** [reply]
Ciao @next731 
perfetto, restiamo a disposizione per ulteriori domande.

Buona giornata
Mauro
